### PR TITLE
Feature 89507 react app  leaderboard component

### PR DESCRIPTION
**Pull request by NataliaMilgevskaya**
Leaderboard component fetches and displays a leaderboard of users based on the total amount of "Zap" (Lightning Network payments or points) each user has accumulated in their private wallets. The component ranks the users according to the total amount of Satoshis (Sats) in descending order and provides the following features:

Key Features:
Fetching Data from an API:

The component fetches user data from the getUsers API and retrieves private wallet information for each user using getUserWallets.
It then fetches the transactions for each private wallet using the fetchWalletTransactions function.
Transaction Summation and Ranking:

The component collects all transactions from the private wallets of users.
It calculates the total amount of Satoshis (Sats) for each user by summing up the transaction amounts (after converting them from millisatoshis to satoshis).
It then ranks the users in descending order based on their total amount of Satoshis.
The ranks are displayed in the leaderboard, with the top 3 ranked users being visually distinguished using specific styles and icons.
Dynamic Sorting:

The leaderboard can be sorted by the total Zap amount in ascending or descending order by clicking the "Zap Amount" header.
The sort direction is toggled each time the header is clicked, and the appropriate sort icon (AscendingIcon or DescendingIcon) is shown.
Visual Representation:

Each row of the leaderboard shows:
The user's rank (with special styling and icons for ranks 1, 2, and 3).
The user's display name.
The total Zap amount (formatted using a number format).
A ZapIcon after the Zap amount.
Users are visually distinguished based on their ranks:
Rank 1: Gold-colored and the first place icon.
Rank 2: Silver-colored and the second place icon.
Rank 3: Bronze-colored and the third place icon.
All other ranks use a default style and icon.
Error Handling and Loading State:

A loading state is shown while the data is being fetched.
If there is an error during data fetching, it is displayed to the user with a relevant error message.
How the Component Works:
Data Fetching:

Upon mounting or when the timestamp prop changes, the component triggers the fetchUsersAndPrivateWalletTransactions function to load user data and wallet transactions.
For each user, their private wallet is identified, and all transactions from that wallet are fetched. If a timestamp prop is provided, the data will be re-fetched when it changes (though it is optional).
Transaction Aggregation:

Once all transactions are retrieved, they are aggregated for each user. The amount for each transaction is summed to give the total Zap amount (in Satoshis) for each user.
Ranking:

The users are ranked based on their total Zap amount. The users with the highest amounts are displayed at the top of the leaderboard.
The component assigns ranks starting from 1 for the user with the highest total Zap amount.
Sorting:

Users can sort the leaderboard in ascending or descending order by Zap amount. Clicking the "Zap Amount" header toggles the sorting order, and the leaderboard is updated accordingly.
Key Functions:
useEffect Hook:

Triggers the data fetching when the component mounts or when the timestamp prop changes.
Data Fetching Functions:

getUsers(): Fetches the list of users.
getUserWallets(): Fetches the wallets associated with each user.
fetchWalletTransactions(): Fetches the transactions for a specific wallet.
Dynamic Rendering:

The component conditionally renders based on the loading and error states. If loading is in progress, it displays a loading message. If there is an error, it shows the error message.
After the data is loaded, the leaderboard is displayed with users' rank, name, and Zap amount.
Props:
timestamp (optional): A timestamp can be passed as a prop to the component. When the timestamp changes, the component will re-fetch the data based on the new time period. This is useful for filtering transactions within a specific timeframe.
Styling:
The component is styled using the Leaderboard.module.css file. It includes:

Special icons and colors for the top 3 ranks.
Dynamic styles for ascending and descending sorting.
Responsive layout for the leaderboard with a clean and clear display of ranks and Zap amounts.

### CODE REVIEW

**Comment by BenGWeeks** [comment]
Tell ChatGPT / Copilot to not be so verbose on the PR description ;-)

**Comment by BenGWeeks** [comment]
Love the video in the PR!!!

File: tabs/src/services/lnbitsServiceLocal.ts [hunk 1]
```diff
@@ -570,6 +570,29 @@ const getWalletZapsSince = async (
   }
 };
 
+const checkWalletExists = async (
```
Comment (BenGWeeks) [inline tabs/src/services/lnbitsServiceLocal.ts:573]: This should be removed, I don't think it's used, and doesn't call LNBits so wouldn't belong in here anyway.

File: tabs/src/components/FooterComponent.tsx [hunk 2]
```diff
@@ -5,6 +5,7 @@ import styles from './FooterComponent.module.css';
 const FooterComponent: React.FC = () => {
   return (
     <footer className={styles.footer}>
+      <Link to="/">Home</Link>
```
Comment (BenGWeeks) [inline tabs/src/components/FooterComponent.tsx:8]: Duplicate?

File: tabs/src/services/lnbitsServiceLocal.ts [hunk 3]
```diff
@@ -740,6 +763,68 @@ const getNostrRewards = async (adminKey: string, stallId: string): Promise<Nostr
   }
 };
 
+interface Transaction {
```
Comment (BenGWeeks) [inline tabs/src/services/lnbitsServiceLocal.ts:766]: This should be in `global.d.ts` (check no existing type for this).

File: tabs/src/services/lnbitsServiceLocal.ts [hunk 4]
```diff
@@ -740,6 +763,68 @@ const getNostrRewards = async (adminKey: string, stallId: string): Promise<Nostr
...
+  checking_id: string;
+  pending: boolean;
+  amount: number; // in millisatoshis (msats)
+  fee: number;
+  memo: string;
+  time: number; // Unix timestamp
+}
+
+interface WalletTransaction {
```
Comment (BenGWeeks) [inline tabs/src/services/lnbitsServiceLocal.ts:775]: What is the difference between `Transaction` and `WalletTransaction`?

File: tabs/src/types/global.d.ts [hunk 5]
```diff
@@ -20,6 +20,7 @@ interface ZapRewards {
 }
 
 interface Wallet {
+  [x: string]: string;
```
Comment (BenGWeeks) [inline tabs/src/types/global.d.ts:23]: What does this do?

**Review by BenGWeeks — CHANGES_REQUESTED** [review]

File: tabs/src/services/lnbitsServiceLocal.ts (hunk 4 above)
Comment (NataliaMilgevskaya) [reply tabs/src/services/lnbitsServiceLocal.ts:775]: The main difference between Transaction and WalletTransaction is how they structure the data and their purposes within the context:

1. Transaction:
This interface represents a single transaction in a wallet.
Each Transaction contains details about a specific transaction such as the amount, fee, memo, timestamp (time), and whether it's pending.

**interface Transaction {
  checking_id: string;
  pending: boolean;
  amount: number;  // in millisatoshis (msats)
  fee: number;
  memo: string;
  time: number;  // Unix timestamp
}**

A Transaction represents a single record of money being sent or received in a wallet.

2. WalletTransaction:
This interface is a container for multiple transactions belonging to a specific wallet.
It groups multiple Transaction objects under a wallet_id (which identifies the wallet they belong to).

**interface WalletTransaction {
  wallet_id: string;
  transactions: Transaction[];
}**
A WalletTransaction groups all the Transaction records for a specific wallet and allows you to associate a list of transactions with a particular wallet.

I would like to leave it as is

**Comment by BenGWeeks** [comment]
OK, in that case you should add “transactions” as a property of Wallet and remove WalletTransactions.

Sent from Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by BenGWeeks** [comment]
Let’s leave as-is for now though. Would need more consoderation.

Sent from Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Pull request merged**